- added features: annotate info documents.

cage2 opened the pull request:
Hello!

this is my attempt to address #48 
there are some points to investigate in my opinion, in details:

- when an info node is annotated and the user try to save the file emacs will ask a file name where dumps the info buffer and this package annotate that saved file, if just 'annnotate-save-annotation' is called the correct file is annotated, though. i am not sure this is an actual bug, this behaviour seems reasonably correct to me; please @randomwangran, can you tell us it this is the same way you used your patched version of this program?

- the function 'annotate-info-select-fn' is a bit slow it is not difficult to understand why ;-);

- on my system if i do not add the test '(when (buffer-file-name)' in the hook called before a buffer's content is changed a weird error complaining about buffer encoding pop up, i suspect there is a better way to get rid of this error but i had no better idea so far.

Bye!
C.

Commit:
- added features: annotate info documents.

bastibe commented on annotate.el:
```diff
@@ -228,30 +228,37 @@ it is called any time the buffer content is changed (so, for
-           ;; delete overlay if there is no more annotated text
-           (when (annotate-annotated-text-empty-p overlay)
-             (delete-overlay overlay))))))))
+  (when (buffer-file-name)
```
We should investigate why this is necessary, as this effectively breaks annotate-mode if a file without a name is being edited.

Could you tell us the error message you are getting if this is left in?

Commit:
- removed check for  buffer file name in funcion called  just before a
  buffer is going to change;
- added check  in 'annotate--remove-annotation-property' to test  if a
  buffer is empty, if it does skip the rest of that function's code.

Commit:
- try to guess  the actual file name of an  info document when showing
  text sippet in summary window.

cage2 commented:
On Wed, Nov 13, 2019 at 12:00:00AM -0800, Bastian Bechtold wrote:

Hello Bastian!

[...]

> +  (when (buffer-file-name)
>

> We should  investigate why  this is  necessary, as  this effectively
> breaks annotate-mode if a file without a name is being edited.

Well      seems       that      i      found       the      "culprit":
`annotate--remove-annotation-property`.

What i think  i have understood so  far: when an info  node is loaded,
except when the root  node is loaded the first time  (i mean when `M-x
info` is called),  the *info* buffer is emptied and  filled again whit
the new  content's node.  This will  trigger `before-change-functions`
and, therefore, `annotate-before-change-fn`.   This function, in turn,
will call `annotate--remove-annotation-property` just after the buffer
has  been emptied.  The problem  is that,  for some  reason i  can not
understand,  the annotation  overlays are  still there  (in the  empty
buffer) from point 1  to point 1 (not a typo) :-/ in  a buffer of size
0...

so, unsurprisingly, this form fails: :-(

``` lisp
(remove-text-properties (point) (1+ (point)) '(display nil)))
```

So i have just removed the test for filename and added a test to check
that we are  not trying to remove  a string property in  a region that
does not exists because the buffer is empty.

I wonder why annotations exists at all in an empty buffer, though. :(

There  was  a problem  with  info  filename  that prevented  the  text
snippet, in  summary window, to shown  at all with an  annotated info
file. I fixed it and now the text appears but is all messed up...

There is still work to do i guess. :)

Bye!
C.

bastibe commented:
Fascinating! So, annotations have two components: Overlays (the red underlines), and a display property at the end of the line that shows the annotation. It sounds like the display property remains, even though the overlay is correctly deleted. How strange.

cage2 commented:
On Thu, Nov 14, 2019 at 12:56:45AM -0800, Bastian Bechtold wrote:

> Fascinating! So, annotations have  two components: Overlays (the red
> underlines), and  a display  property at  the end  of the  line that
> shows the annotation.  It sounds like the  display property remains,
> even though the overlay is correctly deleted. How strange.

yes it is! :)

Turned  out that  the problem  was more  difficult than  expected, but
difficults are parts of the fun i guess! ;-D

Bye!
C.

Commit:
- rewritten  filename juggling  to make  it  works with  info file  in
  summary window;
- made summary window read only;
- other minor fixes/improvements.

bastibe commented:
As the complications of this pull request expand, I think it would be useful to clearly mark info-related code paths as such. Could you add comments and denote functions as `annotate-info-...` to make this clear?

cage2 commented:
On Fri, Nov 15, 2019 at 12:17:50AM -0800, Bastian Bechtold wrote:

Hi Bastian!!

> As the complications  of this pull request expand, I  think it would
> be useful to clearly mark info-related code paths as such. Could you
> add  comments and  denote functions  as `annotate-info-...`  to make
> this clear?

Yes, of course!

Bye!
C.

Commit:
- made recongnizable code that refers to info file annotation (1/3)
  annotate-valid-info-extensions -> annotate-info-valid-file-extensions

Commit:
- made recongnizable code that refers to info file annotation. (2/3)
  added 'annotate-info-actual-filename'

Commit:
- made recongnizable code that refers to info file annotation. (3/3)
  improved comments

cage2 commented:
Hi bastibe!!

> 
> As the complications  of this pull request expand, I  think it would
> be useful to clearly mark info-related code paths as such. Could you
> add  comments and  denote functions  as `annotate-info-...`  to make
> this clear?

Is it OK now?

Bye!
C.

Commit:
- made text snippet in summary window, for info files, less messy.

Commit:
- placed the cursor in the right place (on the annotated text) when an
  info file is reached from the annotation's summary window.

bastibe commented:
Cool, this looks good! merge when you're ready.

Commit:
- using 'annotate-with-inhibit-modification-hooks'.

cage2 commented:
Hi!

Very well, I just committed  a simple modification with 
https://github.com/bastibe/annotate.el/pull/49/commits/75d291a69fb4e10acbafdc7c7f11e4046cec71c7, so tomorrow i will merge that branch into master and close the issue #48.

Bye!!!
C.

cage2 commented:
Ok, merged! Thank you for your help and support bastibe! :)

bastibe commented:
Thank you for all your hard work!

cage2 commented:
On Wed, Nov 20, 2019 at 03:07:04AM -0800, Bastian Bechtold wrote:
> Thank you for all your hard work!

You're welcome! :)

Bye!
C.

randomwangran commented:
I was unable to access my productive computer last month, so I
could not respond to you.

I just did some tests and I think this feature has been
successfully implemented.

Thank you cage2 and bastibe.

cage2 commented:
On Wed, Nov 27, 2019 at 06:52:01AM -0800, randomwangran wrote:

Hi!

> I was unable to access my productive computer last month, so I
> could not respond to you.

No problem!

> I just did some tests and I think this feature has been
> successfully implemented.
>
> Thank you cage2 and bastibe.

Happy to know that everything seems to works fine!

Enjoy this  package, and  thanks for your  suggestion to  improve this
software!

Bye!
C.
